// nativescript-telerik-ui/sidedrawer/sidedrawer.ios.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var commonModule = require("./sidedrawer-common");
var viewModule = require("ui/core/view");
var contentView = require("ui/content-view");
var utils = require("utils/utils");
require("utils/module-merge").merge(commonModule, exports);
; //todo: consider usage of this type for custom styling
////////////////////////////////////////////////
var RadSideDrawer = (function (_super) {
    __extends(RadSideDrawer, _super);
    function RadSideDrawer() {
        _super.call(this);
        this._mainContentHost = new contentView.ContentView();
        this._drawerContentHost = new contentView.ContentView();
        this._ios = TKSideDrawerView.alloc().initWithFrameMainView(UIScreen.mainScreen().bounds, this._mainContentHost.ios);
        this._ios.defaultSideDrawer.content = this._drawerContentHost.ios;
        this._nativeDelegate = TKSideDrawerDelegateImpl.new().initWithOwner(this);
        this._ios.defaultSideDrawer.width = this.drawerContentSize;
        this._ios.defaultSideDrawer.style.blurType = 0;
        this._ios.defaultSideDrawer.headerView = null;
        this._ios.defaultSideDrawer.footerView = null;
        this._addView(this._mainContentHost);
        this._addView(this._drawerContentHost);
        this._ios.defaultSideDrawer.delegate = this._nativeDelegate;
    }
    Object.defineProperty(RadSideDrawer.prototype, "ios", {
        get: function () {
            return this._ios;
        },
        enumerable: true,
        configurable: true
    });
    //data changed event handlers
    RadSideDrawer.prototype._onDrawerLocationChanged = function (eventData) {
        var newLocation = eventData.newValue;
        if (!eventData.newValue) {
            return;
        }
        this.setDrawerLocation(newLocation);
    };
    RadSideDrawer.prototype._onGesturesEnabledChanged = function (eventData) {
        var newValue = eventData.newValue;
        this.ios.defaultSideDrawer.allowGestures = newValue;
    };
    RadSideDrawer.prototype.setDrawerLocation = function (newLocation) {
        var newLocationToLower = newLocation.toLocaleLowerCase();
        switch (newLocationToLower) {
            case commonModule.SideDrawerLocation.Left.toLocaleLowerCase():
                this._ios.defaultSideDrawer.position = TKSideDrawerPosition.TKSideDrawerPositionLeft;
                break;
            case commonModule.SideDrawerLocation.Right.toLocaleLowerCase():
                this._ios.defaultSideDrawer.position = TKSideDrawerPosition.TKSideDrawerPositionRight;
                break;
            case commonModule.SideDrawerLocation.Top.toLocaleLowerCase():
                this._ios.defaultSideDrawer.position = TKSideDrawerPosition.TKSideDrawerPositionTop;
                break;
            case commonModule.SideDrawerLocation.Bottom.toLocaleLowerCase():
                this._ios.defaultSideDrawer.position = TKSideDrawerPosition.TKSideDrawerPositionBottom;
                break;
        }
    };
    RadSideDrawer.prototype._onDrawerContentSizeChanged = function (eventData) {
        var value = eventData.newValue;
        this._ios.defaultSideDrawer.width = value;
    };
    RadSideDrawer.prototype._onDrawerTransitionChanged = function (eventData) {
        var value = eventData.newValue;
        this._ios.defaultSideDrawer.transition = value.getNativeContent();
    };
    RadSideDrawer.prototype._onMainContentChanged = function (eventData) {
        var drawer = eventData.object;
        var newContent = eventData.newValue;
        if (newContent instanceof viewModule.View) {
            drawer._mainContentHost.content = newContent;
        }
    };
    RadSideDrawer.prototype._onDrawerContentChanged = function (eventData) {
        var drawer = eventData.object;
        var newContent = eventData.newValue;
        if (newContent instanceof viewModule.View) {
            drawer._drawerContentHost.content = newContent;
        }
    };
    Object.defineProperty(RadSideDrawer.prototype, "_nativeView", {
        get: function () {
            return this._ios;
        },
        enumerable: true,
        configurable: true
    });
    RadSideDrawer.prototype.closeDrawer = function () {
        if (this._ios) {
            this._ios.defaultSideDrawer.dismiss();
            _super.prototype.closeDrawer.call(this);
        }
    };
    RadSideDrawer.prototype.showDrawer = function () {
        if (this._ios) {
            this._ios.defaultSideDrawer.show();
            _super.prototype.showDrawer.call(this);
        }
    };
    // public onUnloaded() {
    //     console.log("on unloaded");
    //     // this._removeView(this._mainContentHost);
    //     // this._removeView(this._drawerContentHost);
    //     super.onUnloaded();
    // }
    RadSideDrawer.prototype.onLayout = function (left, top, right, bottom) {
        var width = right - left;
        var height = bottom - top;
        var pos = this._ios.defaultSideDrawer.position;
        if (pos == TKSideDrawerPosition.TKSideDrawerPositionTop || pos == TKSideDrawerPosition.TKSideDrawerPositionBottom) {
            this._drawerContentHost.layout(0, 0, width, this.drawerContentSize);
        }
        else {
            this._drawerContentHost.layout(0, 0, this.drawerContentSize, height);
        }
        this._mainContentHost.layout(0, 0, width, height);
    };
    RadSideDrawer.prototype.onMeasure = function (widthMeasureSpec, heightMeasureSpec) {
        var pos = this._ios.defaultSideDrawer.position;
        if (pos == TKSideDrawerPosition.TKSideDrawerPositionTop || pos == TKSideDrawerPosition.TKSideDrawerPositionBottom) {
            viewModule.View.measureChild(this, this._drawerContentHost, widthMeasureSpec, utils.layout.makeMeasureSpec(this.drawerContentSize, utils.layout.EXACTLY));
        }
        else {
            viewModule.View.measureChild(this, this._drawerContentHost, utils.layout.makeMeasureSpec(this.drawerContentSize, utils.layout.EXACTLY), heightMeasureSpec);
        }
        var result = viewModule.View.measureChild(this, this._mainContentHost, widthMeasureSpec, heightMeasureSpec);
        var width = utils.layout.getMeasureSpecSize(widthMeasureSpec);
        var widthMode = utils.layout.getMeasureSpecMode(widthMeasureSpec);
        var height = utils.layout.getMeasureSpecSize(heightMeasureSpec);
        var heightMode = utils.layout.getMeasureSpecMode(heightMeasureSpec);
        var widthAndState = viewModule.View.resolveSizeAndState(result.measuredWidth, width, widthMode, 0);
        var heightAndState = viewModule.View.resolveSizeAndState(result.measuredHeight, height, heightMode, 0);
        this.setMeasuredDimension(widthAndState, heightAndState);
    };
    return RadSideDrawer;
}(commonModule.RadSideDrawer));
exports.RadSideDrawer = RadSideDrawer;
////////////////////////////////////////////////
//              TRANSITIONS
////////////////////////////////////////////////
var FadeTransition = (function (_super) {
    __extends(FadeTransition, _super);
    function FadeTransition() {
        _super.apply(this, arguments);
    }
    FadeTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeFadeIn;
    };
    return FadeTransition;
}(commonModule.DrawerTransitionBase));
exports.FadeTransition = FadeTransition;
var PushTransition = (function (_super) {
    __extends(PushTransition, _super);
    function PushTransition() {
        _super.apply(this, arguments);
    }
    PushTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypePush;
    };
    return PushTransition;
}(commonModule.DrawerTransitionBase));
exports.PushTransition = PushTransition;
var RevealTransition = (function (_super) {
    __extends(RevealTransition, _super);
    function RevealTransition() {
        _super.apply(this, arguments);
    }
    RevealTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeReveal;
    };
    return RevealTransition;
}(commonModule.DrawerTransitionBase));
exports.RevealTransition = RevealTransition;
var ReverseSlideOutTransition = (function (_super) {
    __extends(ReverseSlideOutTransition, _super);
    function ReverseSlideOutTransition() {
        _super.apply(this, arguments);
    }
    ReverseSlideOutTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeReverseSlideOut;
    };
    return ReverseSlideOutTransition;
}(commonModule.DrawerTransitionBase));
exports.ReverseSlideOutTransition = ReverseSlideOutTransition;
var ScaleDownPusherTransition = (function (_super) {
    __extends(ScaleDownPusherTransition, _super);
    function ScaleDownPusherTransition() {
        _super.apply(this, arguments);
    }
    ScaleDownPusherTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeScaleDownPusher;
    };
    return ScaleDownPusherTransition;
}(commonModule.DrawerTransitionBase));
exports.ScaleDownPusherTransition = ScaleDownPusherTransition;
var ScaleUpTransition = (function (_super) {
    __extends(ScaleUpTransition, _super);
    function ScaleUpTransition() {
        _super.apply(this, arguments);
    }
    ScaleUpTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeScaleUp;
    };
    return ScaleUpTransition;
}(commonModule.DrawerTransitionBase));
exports.ScaleUpTransition = ScaleUpTransition;
var SlideAlongTransition = (function (_super) {
    __extends(SlideAlongTransition, _super);
    function SlideAlongTransition() {
        _super.apply(this, arguments);
    }
    SlideAlongTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeSlideAlong;
    };
    return SlideAlongTransition;
}(commonModule.DrawerTransitionBase));
exports.SlideAlongTransition = SlideAlongTransition;
var SlideInOnTopTransition = (function (_super) {
    __extends(SlideInOnTopTransition, _super);
    function SlideInOnTopTransition() {
        _super.apply(this, arguments);
    }
    SlideInOnTopTransition.prototype.getNativeContent = function () {
        return TKSideDrawerTransitionType.TKSideDrawerTransitionTypeSlideInOnTop;
    };
    return SlideInOnTopTransition;
}(commonModule.DrawerTransitionBase));
exports.SlideInOnTopTransition = SlideInOnTopTransition;
////////////////////////////////////////////////
//      Delegate implementation
////////////////////////////////////////////////
var TKSideDrawerDelegateImpl = (function (_super) {
    __extends(TKSideDrawerDelegateImpl, _super);
    function TKSideDrawerDelegateImpl() {
        _super.apply(this, arguments);
    }
    TKSideDrawerDelegateImpl.new = function () {
        return _super.new.call(this);
    };
    TKSideDrawerDelegateImpl.prototype.initWithOwner = function (owner) {
        this._owner = owner;
        return this;
    };
    TKSideDrawerDelegateImpl.prototype.willShowSideDrawer = function (sideDrawer) {
        if (this._owner.hasListeners(commonModule.RadSideDrawer.drawerOpeningEvent)) {
            var args = {
                eventName: commonModule.RadSideDrawer.drawerOpeningEvent,
                object: this._owner,
                returnValue: false
            };
            this._owner.notify(args);
        }
    };
    ;
    TKSideDrawerDelegateImpl.prototype.didShowSideDrawer = function (sideDrawer) {
        if (this._owner.hasListeners(commonModule.RadSideDrawer.drawerOpenedEvent)) {
            var args = {
                eventName: commonModule.RadSideDrawer.drawerOpenedEvent,
                object: this._owner,
            };
            this._owner.notify(args);
        }
    };
    ;
    TKSideDrawerDelegateImpl.prototype.willDismissSideDrawer = function (sideDrawer) {
        if (this._owner.hasListeners(commonModule.RadSideDrawer.drawerClosingEvent)) {
            var args = {
                eventName: commonModule.RadSideDrawer.drawerClosingEvent,
                object: this._owner,
                returnValue: false
            };
            this._owner.notify(args);
        }
    };
    ;
    TKSideDrawerDelegateImpl.prototype.didDismissSideDrawer = function (sideDrawer) {
        if (this._owner.hasListeners(commonModule.RadSideDrawer.drawerClosedEvent)) {
            var args = {
                eventName: commonModule.RadSideDrawer.drawerClosedEvent,
                object: this._owner,
            };
            this._owner.notify(args);
        }
    };
    ;
    TKSideDrawerDelegateImpl.ObjCProtocols = [TKSideDrawerDelegate];
    return TKSideDrawerDelegateImpl;
}(NSObject));
